File: backend/src/main/java/vn/edu/ut/pbms/service/UserService.java
package vn.edu.ut.pbms.service;

import vn.edu.ut.pbms.dto.request.UpdateStatusRequest;
import vn.edu.ut.pbms.dto.request.UserCreateRequest;
import vn.edu.ut.pbms.dto.request.UserUpdateRequest;
import vn.edu.ut.pbms.dto.response.UserListResponse;
import vn.edu.ut.pbms.dto.response.UserResponse;

/**
 * Service interface for User Management business logic.
 * Defines all operations for managing users in the system.
 */
public interface UserService {

    /**
     * Retrieve users with optional keyword search and role filter.
     *
     * @param keyword partial text to match against fullName or phoneNumber
     * @param role    optional role filter (String representation of Role enum)
     * @return UserListResponse containing matched users, total count, and message
     */
    UserListResponse getUsers(String keyword, String role);

    /**
     * Create a new user in the system.
     * Validates unique constraints on username and phoneNumber.
     * Hashes the raw password using BCrypt.
     * Sets default status to ACTIVE.
     *
     * @param request the user creation payload
     * @return the created user as a response DTO
     */
    UserResponse createUser(UserCreateRequest request);

    /**
     * Update an existing user's profile information.
     * Prevents phone number duplication with other accounts.
     *
     * @param id      the ID of the user to update
     * @param request the user update payload
     * @return the updated user as a response DTO
     */
    UserResponse updateUser(long id, UserUpdateRequest request);

    /**
     * Update a user's status (ACTIVE/INACTIVE).
     * Safely parses the status string to the UserStatus enum.
     *
     * @param id      the ID of the user to update
     * @param request the status update payload
     * @return the updated user as a response DTO
     */
    UserResponse updateStatus(long id, UpdateStatusRequest request);
}

File: frontend/src/services/user.service.ts
import axiosClient from "@/lib/axios-client";
import {
  UserResponse,
  UserListResponse,
  UserCreatePayload,
  UserUpdatePayload,
} from "@/types/user.type";

const BASE_PATH = "/users";

export const getUsers = async (
  keyword?: string,
  role?: string
): Promise<UserListResponse> => {
  const params: Record<string, string> = {};
  if (keyword) params.keyword = keyword;
  if (role) params.role = role;

  const response = await axiosClient.get<UserListResponse>(BASE_PATH, {
    params,
  });
  return response.data;
};

export const createUser = async (
  data: UserCreatePayload
): Promise<UserResponse> => {
  const response = await axiosClient.post<UserResponse>(BASE_PATH, data);
  return response.data;
};

export const updateUser = async (
  id: number,
  data: UserUpdatePayload
): Promise<UserResponse> => {
  const response = await axiosClient.put<UserResponse>(`${BASE_PATH}/${id}`, data);
  return response.data;
};

export const updateUserStatus = async (
  id: number,
  status: string
): Promise<UserResponse> => {
  const response = await axiosClient.patch<UserResponse>(`${BASE_PATH}/${id}/status`, {
    status,
  });
  return response.data;
};
